### DISCUSSION TITLE
Improving the UX of `dependsOn` resolution error

### DISCUSSION THREAD

**Discussion by kubukoz** [sbt/sbt#7145] [Ideas]
Hi!

I remembered about this problem I had a long time ago, which still seems to affect users.

When you introduce a `dependsOn` between projects that don't exactly have matching `crossScalaVersions` values, you may see a resolution error like this:

```scala
[error] (main / update) sbt.librarymanagement.ResolveException: Error downloading core:core_3:0.1.0-SNAPSHOT
[error]   Not found
[error]   Not found
[error]   not found: /Users/kubukoz/.ivy2/local/core/core_3/0.1.0-SNAPSHOT/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/core/core_3/0.1.0-SNAPSHOT/core_3-0.1.0-SNAPSHOT.pom
[error] Total time: 2 s, completed Feb 5, 2023, 6:07:17 PM
```

I feel like this may be a case of a leaky abstraction - the first thing I'd think seeing this for the first time is:

> why is Maven/ivy involved in resolving a dependency inside the same project?

I may know this now, but it's still something a user would not expect, and it doesn't tell them what they did wrong or how to fix the problem.

I would like to suggest changing this into something that tells you more about the origin of the problem instead of the symptom: for example, it'd tell you that `core` (a module dependency inside the same project) is missing a variant with the currently used Scala version.

To reproduce:

```scala
val core = project
  .settings(
    crossScalaVersions := Seq("2.13.10")
  )

val main = project
  .settings(
    crossScalaVersions := Seq("2.13.10", "3.2.2")
  )
  .dependsOn(core)
```

1. run sbt (1.8.2)
2. `++3.2.2`
3. `main/compile`

Let me know what you think!

**Comment by eed3si9n** [comment]
Thanks for bringing this up. If we can detect that there's a `ResolveException` and also perform some subproject dependency check, it sounds like a good idea to point users towards the right direction.
